trust-quorum/src/crypto.rs

```diff
@@ -223,6 +256,193 @@ impl Default for Salt {
     }
 }
 
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
+/// All possibly relevant __encrypted__ rack secrets for _prior_ committed
+/// configurations
+pub struct EncryptedRackSecrets {
+    /// A random value used to derive the key to encrypt the rack secrets for
+    /// prior committed epochs.
+    salt: Salt,
+    data: Box<[u8]>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
+pub enum DecryptionError {
+    // An opaque error indicating decryption failed
+    #[error("Failed to decrypt rack secrets")]
+    Aead,
+
+    // The length of the plaintext is not the correct size and cannot
+    // be decoded.
+    #[error("Plaintext length is invalid")]
+    InvalidLength,
+}
+
+impl From<aead::Error> for DecryptionError {
+    fn from(_: aead::Error) -> Self {
+        DecryptionError::Aead
+    }
+}
+
+impl EncryptedRackSecrets {
+    pub fn new(salt: Salt, data: Box<[u8]>) -> Self {
+        EncryptedRackSecrets { salt, data }
+    }
+
+    pub fn decrypt(
+        &self,
+        rack_id: RackUuid,
+        epoch: Epoch,
+        rack_secret: &ReconstructedRackSecret,
+    ) -> Result<PlaintextRackSecrets, DecryptionError> {
+        let key = derive_encryption_key_for_rack_secrets(
+            rack_id,
+            epoch,
+            self.salt,
+            rack_secret,
+        );
+
+        // This key only encrypts a single plaintext and so a nonce of all zeroes
+        // is all that's required.
+        let nonce = [0u8; CHACHA20POLY1305_NONCE_LEN].into();
+
+        let plaintext =
+            Zeroizing::new(key.decrypt(&nonce, self.data.as_ref())?);
+
+        if plaintext.len() % (SECRET_LEN + EPOCH_LEN) != 0 {
+            return Err(DecryptionError::InvalidLength);
+        }
+
+        let mut plaintext_secrets = PlaintextRackSecrets::new();
+
+        for p in plaintext.chunks_exact(SECRET_LEN + EPOCH_LEN) {
+            // SAFETY: Neither of these unwraps will ever fail as we've checked
+            // the validity of plaintext length above.
+            let epoch =
+                Epoch(u64::from_be_bytes(p[0..EPOCH_LEN].try_into().unwrap()));
+            let secret: ReconstructedRackSecret =
+                p[EPOCH_LEN..].try_into().unwrap();
```

Comment:
It almost seems like this would be cleaner if we could use an intermediate struct in both `encrypt` and `decrypt` and serialize/deserialize that instead of just a raw slice except a) we don't know the full number of secrets and b)  I don't know if that has other cryptographic implications.

Reply:
I started to make this change and I have changed my opinion away from serializing via serde. This requires making the `RackSecretData` `Serialize` or pulling the secret out into a separate private intermediate struct like you suggested. However, that means that anytime we add a field we have to update the intermediate struct to copy it. 

I don't think this actually makes things any simpler. I believe it also adds one more spot where the secret lives in memory because you can't move out of a `SecretBox` as far as I know.